File: src/api/v1/modules/movements/domain/entities/Movement.ts
import { Account } from "../../../accounts/domain/entities/Account";
import { Category } from "../../../catecories/domain/entities/Category";
import { Entity } from "../../../shared/domain/entities/Entity";
import { Tag } from "../../../tags/domain/entities/Tag";

export interface MovementType extends Entity {
  type_id: number;
  description: string;
}

export interface Submovement extends Entity {
  submovement_id: number;
  description: string;
  amount: number;
  subcategory: Category;
  tags: Tag[] | [];
}

export interface Movement extends Entity {
  movement_id: number;
  user_id: number;
  title: string;
  description: string;
  amount: number;
  accounting_date: string;
  type: MovementType;
  category: Category;
  account: Account;
  tags: Tag[] | [];
  submovements: Submovement[] | [];
}

File: madfinancial_app/lib/features/movements/domain/entities/movement.dart
import 'package:equatable/equatable.dart';

import 'account.dart';
import 'category.dart';
import 'movement_type.dart';
import 'submovement.dart';
import 'tag.dart';

class Movement extends Equatable {
  const Movement({
    required this.id,
    required this.userId,
    required this.title,
    required this.description,
    required this.amount,
    required this.accountingDate,
    required this.type,
    required this.category,
    required this.account,
    required this.tags,
    required this.submovements,
  });

  final int id;
  final int userId;
  final String title;
  final String description;
  final double amount;
  final DateTime accountingDate;
  final MovementType type;
  final Category category;
  final Account account;
  final List<Tag> tags;
  final List<Submovement> submovements;

  bool get isIncome => type.isIncome;

  double get signedAmount => isIncome ? amount.abs() : -amount.abs();

  @override
  List<Object?> get props => [
    id,
    userId,
    title,
    description,
    amount,
    accountingDate,
    type,
    category,
    account,
    tags,
    submovements,
  ];
}
